packages/wrangler/src/d1/splitter.ts

```diff
@@ -0,0 +1,161 @@
+	}
+}
+
+function splitSqlIntoStatements(sql: string): string[] {
+	const statements: string[] = [];
+	let str = "";
+	const compoundStatementStack: ((s: string) => boolean)[] = [];
+
+	const iterator = sql[Symbol.iterator]();
+	let next = iterator.next();
+	while (!next.done) {
+		const char = next.value;
+
+		if (compoundStatementStack[0]?.(str + char)) {
+			compoundStatementStack.shift();
+		}
+
+		switch (char) {
+			case `'`:
+			case `"`:
+			case "`":
+				str += char + consumeUntilMarker(iterator, char);
+				break;
+			case `$`: {
+				const dollarQuote =
+					"$" + consumeWhile(iterator, isDollarQuoteIdentifier);
+				str += dollarQuote;
+				if (dollarQuote.endsWith("$")) {
+					str += consumeUntilMarker(iterator, dollarQuote);
+				}
+				break;
+			}
+			case `-`:
+				str += char;
+				next = iterator.next();
+				if (!next.done && next.value === "-") {
+					str += next.value + consumeUntilMarker(iterator, "\n");
+					break;
+				} else {
+					continue;
+				}
+			case `/`:
+				str += char;
+				next = iterator.next();
+				if (!next.done && next.value === "*") {
+					str += next.value + consumeUntilMarker(iterator, "*/");
+					break;
+				} else {
+					continue;
+				}
+			case `;`:
+				if (compoundStatementStack.length === 0) {
+					statements.push(str);
+					str = "";
+				} else {
+					str += char;
+				}
+				break;
+			default:
+				str += char;
+				break;
+		}
+
+		if (isCompoundStatementStart(str)) {
+			compoundStatementStack.unshift(isCompoundStatementEnd);
+		}
+
+		next = iterator.next();
+	}
+	statements.push(str);
+
+	return statements
+		.map((statement) => statement.trim())
+		.filter((statement) => statement.length > 0);
+}
+
+/**
+ * Pulls characters from the string iterator while the predicate remains true.
+ */
+function consumeWhile(
+	iterator: Iterator<string>,
+	predicate: (str: string) => boolean
+) {
+	let next = iterator.next();
+	let str = "";
+	while (!next.done) {
+		str += next.value;
+		if (!predicate(str)) {
+			break;
+		}
+		next = iterator.next();
+	}
+	return str;
+}
+
+/**
+ * Pulls characters from the string iterator until the `endMarker` is found.
+ */
+function consumeUntilMarker(iterator: Iterator<string>, endMarker: string) {
+	return consumeWhile(iterator, (str) => !str.endsWith(endMarker));
+}
+
+/**
+ * Returns true if the `str` ends with a dollar-quoted string marker.
+ * See https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-DOLLAR-QUOTING.
+ */
+function isDollarQuoteIdentifier(str: string) {
+	const lastChar = str.slice(-1);
+	return (
+		// The $ marks the end of the identifier
+		lastChar !== "$" &&
+		// we allow numbers, underscore and letters with diacritical marks
+		(/[0-9_]/i.test(lastChar) ||
+			lastChar.toLowerCase() !== lastChar.toUpperCase())
+	);
+}
+
+/**
+ * Returns true if the `str` ends with a compound statement `BEGIN` marker.
+ */
+function isCompoundStatementStart(str: string) {
+	return /[\s]BEGIN[\s]$/.test(str);
```

Comment:
here and below are the square brackets needed?
```suggestion
	return /\sBEGIN\s$/.test(str);
```
also wouldn't you need to add a `+` so that multiple spaces can also be handled?
```suggestion
	return /\s+BEGIN\s+$/.test(str);
```

Reply:
We only care that there is at least one whitespace either side to prevent a false positive match, e.g. `BEGINNING`.